### PR TITLE
Implement tslint enum-naming to enforce PascalCase on enum members

### PR DESCRIPTION
**Pull request by xianny**
## Description

Implements tslint rule to enforce PascalCase for all enum members. Also implements a testing harness.

See updated CHANGELOGS for list of exported enums affected.

## Testing instructions

run `yarn clean && yarn build && yarn test`
## Types of changes

 

 * Breaking change (fix or feature that would cause existing functionality to change) 

## Checklist:

*   [x] Prefix PR title with `[WIP]` if necessary.
*   [x] Add tests to cover changes as needed.
*   [x] Update documentation as needed.
*   [x] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/tslint-config/tslint.json [hunk 1]
```diff
@@ -25,6 +25,7 @@
         "curly": true,
         "custom-no-magic-numbers": [true, 0, 1, 2, 3, -1],
         "encoding": true,
+        "enum-snake-case": false,
```
Comment (albrow) [inline packages/tslint-config/tslint.json]: Did you mean to set this to `true` to enable the linter or are you planning to do that in a separate PR?

File: packages/tslint-config/rules/enumSnakeCaseRule.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        let fix = toSnakeCase(key).toUpperCase();
+        if (value !== undefined) {
+            fix += ` = ${value}`;
+        }
+        return new Lint.Replacement(node.getStart(ctx.sourceFile), node.getWidth(ctx.sourceFile), fix);
+    }
+}
+
+function isSnakeCase(s: string): boolean {
+    const regex = /^[A-Z\d]+(_{1}[A-Z\d]+)*$/g;
```
Comment (albrow) [inline packages/tslint-config/rules/enumSnakeCaseRule.ts]: nit: You could define `regex` outside of the function scope so it can be more efficiently re-used.

File: packages/tslint-config/rules/enumSnakeCaseRule.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            return ctx.addFailureAtNode(node, Rule.FAILURE_STRING, getFix(node));
+        }
+        // Continue recursion into the AST by calling function `cb` for every child of the current node.
+        return ts.forEachChild(node, cb);
+    }
+
+    function getFix(node: ts.Node): Lint.Replacement {
+        const [key, value] = node
+            .getText(ctx.sourceFile)
+            .split('=')
```
Comment (albrow) [inline packages/tslint-config/rules/enumSnakeCaseRule.ts]: Tthis fails on some edge cases where `value` also contains one or more `=`.

For example:

```ts
export enum Operators {
     ASSIGN = '=',
     EQUALS = '==',
     TRIPLE_EQUALS = '===',
}
```

You can use this test to confirm:

```ts
    it('testing fixer with equals sign in value example', () => {
        const src = `enum Operators { Assign = '=', Equals = '==', TripleEquals = '===' }`;
        const expected = `enum Operators { ASSIGN = '=', EQUALS = '==', TRIPLE_EQUALS = '===' }`;
        const actual = getFixedResult(src, rule);
        const result = helper(src, rule);
        assert.equal(result.errorCount, 3); // tslint:disable-line:custom-no-magic-numbers
        assert.equal(actual, expected);
    });
```

The TypeScript parser has likely already sorted this out when constructing the AST. Is there a way for us to take advantage of the existing AST structure instead of trying to do a bit of manual parsing?

File: packages/tslint-config/tslint.json (hunk 1 above)
Comment (xianny) [reply packages/tslint-config/tslint.json]: I meant to leave it as false for this commit. I initially had it as true and started renaming everything, then had second thoughts because it seemed like it could be a big breaking case for small practical benefit. Do you think I should just go ahead and rename everything?

Comment (LogvinovLeon) [reply packages/tslint-config/tslint.json]: I will do that in a separate PR. Maybe package by package.

File: packages/tslint-config/rules/enumSnakeCaseRule.ts (hunk 2 above)
Comment (LogvinovLeon) [reply packages/tslint-config/rules/enumSnakeCaseRule.ts]: What's the point of `{1}`. Can't you just skip it?

```suggestion
    const capitalAlphaNum = '[A-Z\d]+';
    const regex = new Regex('^${capitalAlphaNum}(_${capitalAlphaNum})*$');
```

Feel free to ignore. I'm actually not sure if this is more readable. Regexes with template strings look weird.

**Comment by LogvinovLeon** [comment]
Thanks for writing our first linter rule tests. You're the boss. Approved (assuming @albrow feedback is fixed).

File: packages/tslint-config/tslint.json (hunk 1 above)
Comment (fabioberger) [reply packages/tslint-config/tslint.json]: I would just do it. We need to bite the bullet and I don't expect there to be a TON on places where we haven't been following this rule, right?

**Comment by xianny** [comment]
Hey @fabioberger @LogvinovLeon , @albrow and I actually discussed on Slack whether we should turn this rule on or not and felt like we should punt on this. We were concerned that renaming a bunch of enums would potentially break backwards compatibility on any external packages using exported members, and it's a pretty big cost just for naming style. Also, there are a few style guides that actually recommend using PascalCase for enums:
screen shot 2019-01-07 at 9 49 47 am

Looking at the linter results, most of our enums are already in PascalCase. Maybe I could rewrite the rule to enforce PascalCase? What was the original reasoning for changing to SCREAMING_SNAKE_CASE?

**Comment by fabioberger** [comment]
Hm, I would like us to enforce some sort of convention here. It'll be less headache for everyone involved long-term. Let's discuss.

**Comment by xianny** [comment]
I've updated to PascalCase as discussed. I think these are the only linter errors but I'm not sure if lerna is outputting lint errors for all files.

screen shot 2019-01-07 at 7 56 27 pm
@LogvinovLeon can you take a look? It's things like `ERC20` vs `Erc20` etc. My opinion is that we should allow acronyms, but it's impossible to distinguish them from ALLCAPS so we either ignore ALLCAPS or manually disable the rule.

**Comment by LogvinovLeon** [comment]
What about listing known acronyms. As a config for the rule.
```json
{
    "no-enum-snake-case": [true, "ZRX", "ERC20", "ERC721"],
}
```
@xianny

File: packages/instant/src/types.ts [hunk 4]
```diff
@@ -95,7 +95,7 @@ export enum Network {
 
 export enum ZeroExInstantError {
     AssetMetaDataNotAvailable = 'ASSET_META_DATA_NOT_AVAILABLE',
-    InsufficientETH = 'INSUFFICIENT_ETH',
+    InsufficientEth = 'INSUFFICIENT_ETH',
```
Comment (steveklebanoff) [inline packages/instant/src/types.ts]: Can we use `// tslint:disable-line:enum-naming` here, as we use it above to keep ETH as `ETH` instead of `Eth` ? 

Related: I like the @LogvinovLeon 's idea of having a config for known/whitelisted acronyms, but understand that it could be a follow-up
